subprojects/core/src/main/java/org/gradle/execution/plan/ValuedPathHierarchy.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.internal.snapshot.CaseSensitivity;
+import org.gradle.internal.snapshot.ChildMap;
+import org.gradle.internal.snapshot.ChildMapFactory;
+import org.gradle.internal.snapshot.EmptyChildMap;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import javax.annotation.CheckReturnValue;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+/**
+ * A hierarchy of relative paths with attached values.
+ *
+ * This is an immutable data structure.
+ */
+public final class ValuedPathHierarchy<T> {
```

Comment:
Does `Path` add anything here? I think we should use `ValuedHierarchy`.

Or perhaps `VfsHierarchy`, given that the thing is addressed via a `VfsRelativePath`.

Reply:
Renamed to `ValuedVfsHIerarchy`.